compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -834,12 +834,123 @@ private static boolean isDefinedInCurrentModuleInCurrentCompilation(SymbolEnv en
         // may not be a corresponding type definition. So we check by type definition if the type corresponds to a
         // type definition defined in the current compilation.
         BType effectiveType = immutableType.effectiveType;
-        int tag = Types.getReferredType(effectiveType).tag;
+        return requiredTypeDefsExist(env, effectiveType, new HashSet<>());
+    }
 
-        if (tag != TypeTags.TUPLE && tag != TypeTags.UNION && tag != TypeTags.RECORD && tag != TypeTags.OBJECT) {
-            return true;
+    private static boolean hasNamedDefs(BType mutableType) {
+        switch (mutableType.tag) {
+            case TypeTags.RECORD:
+            case TypeTags.OBJECT:
+            case TypeTags.TYPEREFDESC:
+                return true;
+            case TypeTags.UNION:
+                return ((BUnionType) mutableType).isCyclic;
+            case TypeTags.TUPLE:
+                return ((BTupleType) mutableType).isCyclic;
         }
+        return false;
+    }
+
+    private static boolean requiredTypeDefsExist(SymbolEnv env, BType type, Set<BType> visitedTypes) {
+        if (!visitedTypes.add(type)) {
+            return false;
+        }
+
+        BType referredType = Types.getReferredType(type);
+        int tag = referredType.tag;
+        switch (tag) {
+            case TypeTags.ARRAY:
+                return requiredTypeDefsExist(env, ((BArrayType) referredType).eType, visitedTypes);
+            case TypeTags.TUPLE:
+                if (relevantTypeDefExists(env, type)) {
+                    return true;
+                }
+                BTypeSymbol tupleTypeSymbol = type.tsymbol;
+
+                boolean allMatched = true;
+                if (tupleTypeSymbol != null && tupleTypeSymbol.name != null && !tupleTypeSymbol.name.value.isEmpty()) {
+                    ((SelectivelyImmutableReferenceType) referredType).unsetImmutableType();
+                    allMatched = false;
+                }
+
+                BTupleType tupleType = (BTupleType) referredType;
+                for (BType tupleMemberType : tupleType.getTupleTypes()) {
+                    if (!requiredTypeDefsExist(env, tupleMemberType, visitedTypes)) {
+                        allMatched = false;
+                    }
+                }
+
+                BType restType = tupleType.restType;
+                if (restType == null) {
+                    return allMatched;
+                }
+                return allMatched && requiredTypeDefsExist(env, restType, visitedTypes);
+            case TypeTags.MAP:
+                return requiredTypeDefsExist(env, ((BMapType) referredType).constraint, visitedTypes);
+            case TypeTags.RECORD:
+            case TypeTags.OBJECT:
+                boolean relTypeDefExistsForRecordOrObject = relevantTypeDefExists(env, type);
+                if (!relTypeDefExistsForRecordOrObject) {
+                    ((SelectivelyImmutableReferenceType) referredType).unsetImmutableType();
```

Comment:
Can there be a case where we unset the immutable type in module `A` in module `B` and again tries to set the immutable type for the type at module `A` where typedef is already there?

Reply:
We always overwrite in the current module, right?